### PR TITLE
Generate internal cask json v3

### PR DESCRIPTION
**Pull request by apainintheneck**
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/HEAD/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/HEAD/Library/Homebrew/test/PATH_spec.rb).
- [x] Have you successfully run `brew style` with your changes locally?
- [x] Have you successfully run `brew typecheck` with your changes locally?
- [ ] Have you successfully run `brew tests` with your changes locally?

-----

This is a sister PR to https://github.com/Homebrew/brew/pull/16460. It start generating the `internal_cask_v3.json` file in the `brew generate-cask-api` file.

I didn't add any tests because they'll be easier to add later on when working on the cask loading code in the near future. The main thing here is that this doesn't break the `brew generate-cask-api` command which is easy to test.

I took the same approach as the other PR to cache the variations generation so the runtime of this command should be similar to what it was before.

The file size difference is significant but the old file wasn't as large as with formulae. It reduces the size from around 6MB to 4MB.

Related to: https://github.com/Homebrew/brew/issues/16410

### CODE REVIEW

**Comment by apainintheneck** [comment]
Moving back to draft since the discussion in https://github.com/Homebrew/brew/pull/16460 is relevant for this PR and could result in changes to naming among other things.

File: Library/Homebrew/cask/cask.rb [hunk 1]
```diff
@@ -380,10 +381,22 @@ def to_hash_with_variations
...
-      hash["variations"] = variations
+      hash["variations"] = @variations_hash
       hash
     end
 
+    def to_api_hash
+      api_hash = to_hash_with_variations.except(
+        # Included in the top-level of the `internal_cask_v3.json` file
+        # created by the `brew generate-cask-api` command.
+        "old_tokens",
```
Comment (MikeMcQuaid) [inline Library/Homebrew/cask/cask.rb:392]: - Unless I'm mistaken: `token` and `full_token` are always identical, too.
- `url_specs`, `languages` seem to be often empty. 
- stuff like `appcast` is often `null`
- `installed` and `installed_time` and `outdated` are reflective of the machine they are run on and not public API.

**Review by MikeMcQuaid — COMMENTED** [review]
Looks good so far, thanks!

File: Library/Homebrew/cask/cask.rb (hunk 1 above)
Comment (SMillerDev) [reply Library/Homebrew/cask/cask.rb:392]: `appcast` is deprecated, if it's not `null` it should be.

**Comment by apainintheneck** [comment]
Closing because the approach in the formula PR changed significantly which means this will need to be reworked.

**Pull request closed without merging**